index.coffee

```diff
@@ -1,44 +1,99 @@
+EventEmitter = require 'events'
+request = require 'simple-get'
 stream = require 'readable-stream'
 throttle = require 'lodash.throttle'
 
-# use the same module as request.js, to preserve behaviour.
 # Object.assign() doesn't fit, because it copies `undefined`
 # props as well, whereas we'd rather not.
 extend = require 'extend'
 
-module.exports = (requestModule, requestOpts, opts) ->
-	return new ResumableRequest(requestModule, requestOpts, opts)
+module.exports = (opts) ->
+	return new ResumableRequest(opts)
 
 module.exports.defaults = (defaults = {}) ->
-	(requestModule, requestOpts, opts) ->
+	(opts) ->
 		opts = extend {}, defaults, opts
-		return new ResumableRequest(requestModule, requestOpts, opts)
+		return new ResumableRequest(opts)
 
-cloneResponse = (response) ->
-	# creates an `http.IncomingMessage`-like stream, that can also be piped to.
-	s = new stream.PassThrough()
-	[ 'headers', 'httpVersion', 'method', 'rawHeaders', 'statusCode', 'statusMessage' ].forEach (prop) ->
-		s[prop] = response[prop]
-	return s
-
-class ResumableRequest extends stream.Readable
-	constructor: (@requestModule, @requestOpts, opts = {}) ->
+class _ResponseProxy extends stream.Readable
+	constructor: (response, @timeout, @onProgress) ->
+		super()
+		@stream = response
+		[ 'headers', 'httpVersion', 'method', 'rawHeaders', 'statusCode', 'statusMessage' ].forEach (prop) =>
+			this[prop] = response[prop]
+		@_ended = false
+
+	follow: (response) ->
+		if @stream?
+			@stream.resume()
+			@stream.removeListener('readable', @_read)
+		@stream = response
+		@stream.once('readable', @_read)
+
+	_read: =>
+		return if @_ended
+
+		chunk = null
+		chunksRead = 0
+
+		# Read chunks from the source & push them out
+		while (chunk = @stream.read())
+			chunksRead++
+
+			@onProgress(chunk)
+
+			# Avoid pushing out more chunks than the destination can handle
+			if @push(chunk) is false
+				break
+
+		# If no chunks were read, wait until the source becomes readable again
+		if chunksRead is 0
+			@stream.once('readable', @_read)
+
+	_destroy: ->
+		@_ended = true
+		@push(null)
+		if @stream?
+			@stream.resume()
+			@stream.removeListener('readable', @_read)
+
+class ResumableRequest extends EventEmitter
+	constructor: (opts) ->
 		super()
 
+		if not opts.url?
+			throw new Error('You must specify a URL')
+		if opts.method? and opts.method.toLowerCase() isnt 'get'
+			throw new Error('Only GET requests are currently supported')
+
+		defaultOpts = {
+			maxRetries: 10
+			retryInterval: 1000
+			progressInterval: 1000
+		}
+
+		opts = extend(defaultOpts, opts)
+		{ maxRetries, retryInterval, progressInterval, timeout } = opts
```

Comment:
You can just use `@maxRetries` here and remove line 92. And you could use `@timeout` and replace `@requestTimeout` I guess.